File: demo/src/main/java/com/juankysoriano/rainbow/demo/sketch/rainbow/RainbowLineCirclesSketch.java
package com.juankysoriano.rainbow.demo.sketch.rainbow;

import android.view.MotionEvent;
import android.view.ViewGroup;

import com.juankysoriano.rainbow.core.Rainbow;
import com.juankysoriano.rainbow.core.drawing.Modes;
import com.juankysoriano.rainbow.core.drawing.RainbowDrawer;
import com.juankysoriano.rainbow.core.drawing.RainbowDrawer.Precision;
import com.juankysoriano.rainbow.core.event.RainbowInputController;

public class RainbowLineCirclesSketch extends Rainbow implements RainbowInputController.RainbowInteractionListener {

    private RainbowDrawer.PointDetectedListener pointDetectedListener = new RainbowDrawer.PointDetectedListener() {

        @Override
        public void onPointDetected(float px, float py, float x, float y) {
            getRainbowDrawer().stroke(0, 30);
            getRainbowDrawer().fill(0, 0);
            getRainbowDrawer().ellipseMode(Modes.Draw.CENTER);
            getRainbowDrawer().ellipse(x, y, 200, 200);
        }

    };

    public RainbowLineCirclesSketch(ViewGroup viewGroup) {
        super(viewGroup);
    }

    @Override
    public void onSketchSetup() {
        stepRate(500);
        getRainbowDrawer().background(255);
        getRainbowDrawer().smooth();
        getRainbowDrawer().noFill();
    }

    @Override
    public void onDrawingStart() {
        getRainbowInputController().attach(this);
    }

    @Override
    public void onDrawingStop() {
        getRainbowInputController().detach();
    }

    @Override
    public void onSketchDestroy() {
        pointDetectedListener = null;
    }

    @Override
    public void onSketchTouched(MotionEvent event, RainbowDrawer rainbowDrawer) {
        //no-op
    }

    @Override
    public void onSketchReleased(MotionEvent event, RainbowDrawer rainbowDrawer) {
        //no-op
    }

    @Override
    public void onFingerDragged(MotionEvent event, RainbowDrawer rainbowDrawer) {
        RainbowInputController rainbowInputController = getRainbowInputController();
        float x = rainbowInputController.getX();
        float y = rainbowInputController.getY();
        float oldX = rainbowInputController.getPreviousX();
        float oldY = rainbowInputController.getPreviousY();
        getRainbowDrawer().exploreLine(oldX, oldY, x, y, Precision.VERY_HIGH, pointDetectedListener);
    }

    @Override
    public void onMotionEvent(MotionEvent event, RainbowDrawer rainbowDrawer) {

    }
}

File: demo/src/main/java/com/juankysoriano/rainbow/demo/sketch/rainbow/RainbowLineCirclesSketch.kt
package com.juankysoriano.rainbow.demo.sketch.rainbow

import android.view.MotionEvent
import android.view.ViewGroup

import com.juankysoriano.rainbow.core.Rainbow
import com.juankysoriano.rainbow.core.drawing.Modes
import com.juankysoriano.rainbow.core.drawing.RainbowDrawer
import com.juankysoriano.rainbow.core.drawing.RainbowDrawer.Precision
import com.juankysoriano.rainbow.core.touch.RainbowInputController

class RainbowLineCirclesSketch(viewGroup: ViewGroup) : Rainbow(viewGroup), RainbowInputController.RainbowInteractionListener {

    override fun onSketchSetup() {
        with(rainbowDrawer) {
            background(255)
            smooth()
            noFill()
        }
    }

    override fun onDrawingStart() = rainbowInputController.attach(this)
    override fun onDrawingStop() = rainbowInputController.detach()

    override fun onSketchDestroy() {
        //no-op
    }

    override fun onSketchTouched(event: MotionEvent) {
        //no-op
    }

    override fun onSketchReleased(event: MotionEvent) {
        //no-op
    }

    override fun onFingerDragged(event: MotionEvent) {
        val rainbowInputController = rainbowInputController
        rainbowDrawer.exploreLine(
                x1 = rainbowInputController.previousX,
                y1 = rainbowInputController.previousY,
                x2 = rainbowInputController.x,
                y2 = rainbowInputController.y,
                precision = Precision.HIGH,
                listener = object : RainbowDrawer.PointDetectedListener {
                    override fun onPointDetected(px: Float, py: Float, x: Float, y: Float) {
                        with(rainbowDrawer) {
                            stroke(0, 30f)
                            fill(0, 0f)
                            ellipseMode(Modes.Draw.DIAMETER)
                            ellipse(x, y, 200f, 200f)
                        }
                    }

                }
        )
    }

    override fun onMotionEvent(event: MotionEvent) {
        //no-op
    }
}
